File: 백준 문제 풀이 모음/python/class 3/16928.py
import sys

N, M = map(int, sys.stdin.readline().split())
dict = {}

for _ in range(N+M) :
    a, b = map(int, sys.stdin.readline().split())
    dict[a] = b

from collections import deque



def bfs (num) :
    queue = deque()
    queue.append([num])

    count = 0
    while queue :
        node = queue.popleft()
        count += 1


        if 100 in node :
            break

        tmp = []
        for i in range(len(node)) :
            for j in range(6) :
                if node[i] + j + 1 not in tmp :
                    if node[i] + j + 1 in dict :
                        if dict[node[i] + j + 1] not in tmp :
                            tmp.append(dict[node[i] + j + 1])
                    else :
                        tmp.append(node[i] + j + 1)
        queue.append(tmp)

    return count

print(bfs(1)-1)
        

File: 백준 문제 풀이 모음/kotlin/class 3/16928.kt


var game = MutableList<Int>(101) { 0 }
var count = 0
fun main() = with(System.`in`.bufferedReader()) {

    val str = readLine().split(' ').toList().map { it.toInt() }
    val n = str[0]
    val m = str[1]

    repeat(n) {
        val input = readLine().split(' ').toList().map { it.toInt() }
        game[input[0]] = input[1]
    }

    repeat(m) {
        val input = readLine().split(' ').toList().map { it.toInt() }
        game[input[0]] = input[1]
    }

//    println(game)
    BFS(1)

}

fun BFS(num: Int) {
    val deque = ArrayDeque<MutableSet<Int>>()
    deque.add(mutableSetOf(num))

    while (deque[0].size != 0) {
//        println(deque)
        val node = deque.removeFirst()
        if (node.contains(100)) {
            println(count)
            break
        }
        val new_list = mutableSetOf<Int>()

        node.forEach {
            for (i in 1..6) {
                if (it + i > 100) continue

                if (game[it + i] == 0) {
                    new_list.add(it + i)
                } else {
                    new_list.add(game[it + i])
                }
            }
        }

        deque.add(new_list)
        count++
    }

}
